src/lib/support/ThreadOperationalDataset.h

```diff
@@ -0,0 +1,255 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <core/CHIPCore.h>
+#include <platform/internal/DeviceNetworkInfo.h>
+
+namespace chip {
+namespace Thread {
+
+class Tlv;
+
+constexpr size_t kChannel_NotSpecified = UINT8_MAX;
+constexpr size_t kPANId_NotSpecified   = UINT16_MAX;
+
+constexpr size_t kSizeNetworkName     = 16;
+constexpr size_t kSizeExtendedPanId   = 8;
+constexpr size_t kSizeMasterKey       = 16;
+constexpr size_t kSizeMeshLocalPrefix = 8;
+constexpr size_t kSizePSKc            = 16;
+
+/**
+ * This class provides methods to manipulate Thread operational dataset.
+ *
+ */
+class OperationalDataset : public DeviceLayer::Internal::ThreadOperationalDataset
```

Comment:
Can we make ThreadOperationalDataset a private data member instead of inherit it and provide an accessory to the pointer.

Reply:
Can we still cast from `ThreadOperationalDataset` in that case? I'd like to keep the ability to cast from the raw data.

Reply:
You can cast an object to its first non-static data member's type if the object fits the constraint std::is_standard_layout, and you can cast such a converted alias back to the object's dynamic type.

However, it would not be legal to cast object whose dynamic type is ThreadOperationalDataset to OperationalDataset. The rule is not symmetric in this way. It does not matter whether ThreadOperationalDataset is the base class or non-static data first member, such a cast breaks aliasing requirements and is undefined behavior. See https://en.cppreference.com/w/cpp/language/reinterpret_cast#Type_aliasing for a list of cases where accessing an object through a type other than its dynamic type is allowed.

Reply:
In the current implementation, `static_cast` is used instead of `reinterpret_cast` from `ThreadOperationalDataset` to `Thread::OperationalDataset`. From https://en.cppreference.com/w/cpp/language/static_cast, such kind of downcast is legal. Should we keep it this way for now?

Reply:
The static_cast is not legal unless the dynamic type is Thread::OperationalDataset

That's what the page means by 

> Such a downcast makes no runtime checks to ensure that the object's runtime type is actually D,

I don't understand why we would add a hierarchy of types and make an API that takes a BaseClass* when it actually needs a DerivedClass* without documenting that fact. It makes no sense.

Reply:
Changed to use ByteSpan instead.